### ISSUE TITLE
Getting wallpaper from Plasma is complicated and prone to errors

### ISSUE THREAD

**Issue by luisbocanegra** [luisbocanegra/kde-material-you-colors#187]
Currently there are two methods to get the current wallpaper, Plasma scripting API and KWin scripting API with the compiled helper.

Plasma scripting method requires wallpaper plugins to actually store the current wallpaper source in the configuration, some do, but most do not (because they don't need to). Also some may require special handling like Picture of the day or store wrong/outdated data, causing us to get the wrong colors.

The fallback screenshot method is more reliable, but also more resource expensive as it requires continuous snapshots and checking whether the image hash changed each time. It's also more complex, requires maintaining both KWin Script + compiled plugin that access the KWin's screenshot plugin. An example of this is the upcoming Plasma 6 KWin changes mentioned in https://github.com/luisbocanegra/kde-material-you-colors/issues/161

For all the problems above I have even considered dropping automatic detection from plasma and have the screenshot method executed manually by spectacle in combination of the show desktop shortcut with just a click on a button. That would be a downgrade but maybe an acceptable one?

**Possible alternative (but still complicated)**

Just recently found out about `Plasmoid.containment.wallpaperGraphicsObject` which is attached to widgets that are placed on the desktop, using `grabToImage` method is possible to store and save an image of what's currently shown as wallpaper.

Additionally to grabbing the image, `wallpaperGraphicsObject` could be recursively scanned for properties like `imagePath, modelImage, color` for some wallpapers where it returns an appropriate source.

The proposed solution is to create an invisible plasmoid to be placed on the desktop, process of adding it to every screen/(activity?) could probably be handled by the backend (Plasma scripting API) or done manually.

Then plasmoids would store the desktop preview to a temporary location (probably scaled down to save resources) and then the backend will make use of them.

It would be necessary to store from which screen/(activity?) the wallpaper is coming from for the backend to know that the current one is.

**Additional context, should this be provided by Plasma?**

Since there are already multiple utilities to set/set wallpaper/icons/colors in plasma, I believe a method to get the actual wallpaper source or even a image snapshot should exist too (currently it only returns their configuration as seen in `qdbus org.kde.plasmashell /PlasmaShell org.kde.PlasmaShell.wallpaper 0`), my plan to raise this idea to KDE developers to see that they think of that (maybe too niche?) feature in Plasma.

**TL;DR**
Looking for a more reliable way to get the wallpaper source or considering dropping automatic detection in favor of a button to do it.

**Comment by Blacksuan19** [comment]
I came here to open an issue about the wallpaper detection not working with wallhaven.cc wallpaper plugin, sometimes it would take a sceenshot which works but most of the time it is using that journalctl command.

after reading this, I can see the issue, the screenshot method is the most reliable currently and decouples wallpaper detection from the plugin, I think an option to only use screenshot method for detecting wallpaper would help with usability for people like me who don't use any of the supported wallpaper plugins until there is a stable API to obtain the current wallpaper.

on another note, plasma also has material-you but it only updates the color scheme colors, it can tint all colors with the extracted colors from wallpaper, perhaps letting plasma handle color extraction and simply taking the values from there would be easier than manual extraction, eliminating the issue of getting wallpaper.

color scheme options
 
![image](https://github.com/luisbocanegra/kde-material-you-colors/assets/10248473/4f49c8c7-d678-4556-9dc1-16df3a75230d)


example color scheme generated from wallpaper


the colors are also applied to the icon pack if supported (breeze does support it, icons shown below are from tela)


![Screenshot_20240227_131510](https://github.com/luisbocanegra/kde-material-you-colors/assets/10248473/06190dbe-151f-43b6-967d-7ccf991ff7c9)

**Comment by luisbocanegra** [comment]
> I came here to open an issue about the wallpaper detection not working with wallhaven.cc wallpaper plugin, sometimes it would take a sceenshot which works but most of the time it is using that journalctl command

It's good to hear it  works.

> after reading this, I can see the issue, the screenshot method is the most reliable currently and decouples wallpaper detection from the plugin, I think an option to only use screenshot method for detecting wallpaper would help with usability for people like me who don't use any of the supported wallpaper plugins until there is a stable API to obtain the current wallpaper.

Sounds like a good idea. The only downside is requiring people who don't use Arch + AUR having to compile the helper but there's no way around that one.

> on another note, plasma also has material-you but it only updates the color scheme colors, it can tint all colors with the extracted colors from wallpaper, perhaps letting plasma handle color extraction and simply taking the values from there would be easier than manual extraction, eliminating the issue of getting wallpaper.
> 
> color scheme options
> 

<details>
<summary>Image</summary>

> ![image](https://private-user-images.githubusercontent.com/10248473/308307225-4f49c8c7-d678-4556-9dc1-16df3a75230d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNjM5MDUsIm5iZiI6MTcwOTA2MzYwNSwicGF0aCI6Ii8xMDI0ODQ3My8zMDgzMDcyMjUtNGY0OWM4YzctZDY3OC00NTU2LTlkYzEtMTZkZjNhNzUyMzBkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDE5NTMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY2YWUyZTE2NjNlOWNjOTQ4YTg1YjlmODgwYWJmOTc0NjgzOWU3YzE2NDllMTU1NTljY2YyYmUyYTRmZmYzZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.BIgKthqh5GJG_hoLVwRtAIp1lE8N7lrpPFYaDzsss6k)

</details>

> 
> example color scheme generated from wallpaper
> 
> the colors are also applied to the icon pack if supported (breeze does support it, icons shown below are from tela)

<details>
<summary>Image</summary>

> ![Screenshot_20240227_131510](https://private-user-images.githubusercontent.com/10248473/308308642-06190dbe-151f-43b6-967d-7ccf991ff7c9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNjM5MDUsIm5iZiI6MTcwOTA2MzYwNSwicGF0aCI6Ii8xMDI0ODQ3My8zMDgzMDg2NDItMDYxOTBkYmUtMTUxZi00M2I2LTk2N2QtN2NjZjk5MWZmN2M5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDE5NTMyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhMzI1MGFlZjgyYTgzZWVhMDMxYjkzZTU3MWVhYjQyZDE2OTAwY2UzOTAyMzJjMmM4YjgzMzY5N2RkNDBiNGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6VXlO_lUIqy0FN6-Nx8EbcsJruzDA42ag8ROk8wqzXA)

</details>

I also thought of adding that as an option, didn't like going from a variable amount of colors returned from the wallpaper to just a single color returned from Plasma API but maybe it's fine for people who don't enable text colors from wallpaper? Shouldn't be too hard to implement.

**Comment by kun-codes** [comment]
offtopic but what plasma style and application style are you using? @Blacksuan19

**Comment by Blacksuan19** [comment]
> Sounds like a good idea. The only downside is requiring people who don't use Arch + AUR having to compile the helper but there's no way around that one.  

you can use github actions to build the helper and upload it as an artifact for popular options like deb or rpm.

> I also thought of adding that as an option, didn't like going from a variable amount of colors returned from the wallpaper to just a single color returned from Plasma API but maybe it's fine for people who don't enable text colors from wallpaper? Shouldn't be too hard to implement.

it's definitely way less colors but its decent enough for themeing purposes.

> offtopic but what plasma style and application style are you using? @Blacksuan19

plasma style is my own version of [colloid-light](https://github.com/vinceliuice/Colloid-kde) that follows system colorscheme
application style is [lightly](https://github.com/Luwx/Lightly)

**Comment by luisbocanegra** [comment]
> > Sounds like a good idea. The only downside is requiring people who don't use Arch + AUR having to compile the helper but there's no way around that one.
> 
> you can use github actions to build the helper and upload it as an artifact for popular options like deb or rpm.

Maybe, previously there were rpm packages built on OBS by a user but they couldn't maintain it anymore and nobody took over.

> > I also thought of adding that as an option, didn't like going from a variable amount of colors returned from the wallpaper to just a single color returned from Plasma API but maybe it's fine for people who don't enable text colors from wallpaper? Shouldn't be too hard to implement.
> 
> it's definitely way less colors but its decent enough for themeing purposes.

True, will add it as an option in a future version

**Comment by luisbocanegra** [comment]
I am aware, but AFAIK there is no way to create a always active non-fullscreen QML to be able to take advantage of `KWinComponents.DesktopBackground`.

**Comment by luisbocanegra** [comment]
> or just a cpp logic plugin without any kcm??

That would be the way to go.

**Comment by saberr26** [comment]
> Nope, at least I couldn't find a way when I was trying to port https://github.com/emvaized/kde-snap-assist as a effect (cant get keyboard presses from a KWin script on wayland)

but a lot of plugins don’t really handle input or fullscreen scenes either (like background-contrast, touch-points, even some of the tool/debug ones). They just pass stuff to KWin. So maybe the problem wasn’t “it can’t be done,” but just that snap-assist needed key events, which scripts don’t get on Wayland
